source/common/websocket/codec.h

```diff
@@ -0,0 +1,168 @@
+                                                  kFrameOpcodeBinary,       kFrameOpcodeClose,
+                                                  kFrameOpcodePing,         kFrameOpcodePong};
+
+// Length of the masking key which is 4 bytes fixed size
+constexpr uint8_t kMaskingKeyLength = 4;
+
+// 16 bit payload length
+constexpr uint8_t kPayloadLength16Bit = 2;
+
+// 64 bit payload length
+constexpr uint8_t kPayloadLength64Bit = 8;
+
+// Wire format (https://datatracker.ietf.org/doc/html/rfc6455#section-5.2)
+// of WebSocket frame:
+//
+//   0                   1                   2                   3
+//   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  +-+-+-+-+-------+-+-------------+-------------------------------+
+//  |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
+//  |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
+//  |N|V|V|V|       |S|             |   (if payload len==126/127)   |
+//  | |1|2|3|       |K|             |                               |
+//  +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
+//  |     Extended payload length continued, if payload len == 127  |
+//  + - - - - - - - - - - - - - - - +-------------------------------+
+//  |                               | Masking-key, if MASK set to 1 |
+//  +-------------------------------+-------------------------------+
+//  | Masking-key (continued)       |          Payload Data         |
+//  +-------------------------------- - - - - - - - - - - - - - - - +
+//  : .... Payload Data continued .... Payload Data continued ..... :
+//  + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+//  | .... Payload Data continued .... Payload Data continued ..... |
+//  +---------------------------------------------------------------+
+
+// In-memory representation of the contents of a WebSocket frame.
+struct Frame {
+  // Indicates that this is the final fragment in a message.
+  bool final_fragment_;
+  // Frame opcode.
+  uint8_t opcode_;
+  // Length of the payload as the number of bytes.
+  uint64_t payload_length_;
+  // The 4 byte fixed size masking key used to mask the payload. Masking/unmasking should be
+  // performed as described in https://datatracker.ietf.org/doc/html/rfc6455#section-5.3
+  absl::optional<uint32_t> masking_key_;
+  // WebSocket payload data (extension data and application data).
+  Buffer::InstancePtr payload_;
+};
+
+// Encoder encodes in memory WebSocket frames into frames in the wire format
+class Encoder {
+public:
+  Encoder() = default;
+
+  // Creates a new Websocket data frame header with the given frame data.
+  // @param frame supplies the frame to be encoded.
+  // @return std::vector<uint8_t> buffer with encoded header data.
+  std::vector<uint8_t> newFrameHeader(const Frame& frame);
+};
+
+// Current state of the frame that is being processed.
+enum class State {
```

Comment:
nit: Since this is basically an implementation detail for Decoder (or DecoderBase currently) consider moving this to be a private inner class of Decoder (or DecoderBase).